pkg/kv/dynamodb/store.go

```diff
@@ -0,0 +1,320 @@
+	return s.params.TableName
+}
+
+// setupKeyValueDatabase setup everything required to enable kv over postgres
+func setupKeyValueDatabase(ctx context.Context, svc *dynamodb.DynamoDB, params *Params) error {
+	// main kv table
+	_, err := svc.CreateTableWithContext(ctx, &dynamodb.CreateTableInput{
+		TableName: aws.String(params.TableName),
+		AttributeDefinitions: []*dynamodb.AttributeDefinition{
+			{
+				AttributeName: aws.String(PartKey),
+				AttributeType: aws.String("S"),
+			},
+			{
+				AttributeName: aws.String(ItemKey),
+				AttributeType: aws.String("S"),
+			},
+		},
+		KeySchema: []*dynamodb.KeySchemaElement{
+			{
+				AttributeName: aws.String(PartKey),
+				KeyType:       aws.String("HASH"),
+			},
+			{
+				AttributeName: aws.String(ItemKey),
+				KeyType:       aws.String("RANGE"),
+			},
+		},
+		ProvisionedThroughput: &dynamodb.ProvisionedThroughput{
+			ReadCapacityUnits:  aws.Int64(ReadCapacityUnits),
+			WriteCapacityUnits: aws.Int64(WriteCapacityUnits),
+		},
+	})
+	if err != nil {
+		if _, ok := err.(*dynamodb.ResourceInUseException); !ok {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *Store) bytesKeyToDynamoKey(key []byte) map[string]*dynamodb.AttributeValue {
+	return map[string]*dynamodb.AttributeValue{
+		PartKey: {
+			S: aws.String(s.getPartitionKey(key)),
+		},
+		ItemKey: {
+			S: aws.String(string(key)),
+		},
+	}
+}
+
+func (s *Store) Get(ctx context.Context, key []byte) (*kv.ValueWithPredicate, error) {
+	if len(key) == 0 {
+		return nil, kv.ErrMissingKey
+	}
+
+	result, err := s.svc.GetItemWithContext(ctx, &dynamodb.GetItemInput{
+		TableName: aws.String(s.params.TableName),
+		Key:       s.bytesKeyToDynamoKey(key),
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	if result.Item == nil {
+		return nil, kv.ErrNotFound
+	}
+
+	var item DynKVItem
+	err = dynamodbattribute.UnmarshalMap(result.Item, &item)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %w", err, kv.ErrOperationFailed)
+	}
+
+	return &kv.ValueWithPredicate{
+		Value:     []byte(item.ItemValue),
+		Predicate: kv.Predicate([]byte(item.ItemValue)),
+	}, nil
+}
+
+func (s *Store) Set(ctx context.Context, key, value []byte) error {
+	return s.setWithOptionalPredicate(ctx, key, value, nil, false)
+}
+
+func (s *Store) SetIf(ctx context.Context, key, value []byte, valuePredicate kv.Predicate) error {
+	return s.setWithOptionalPredicate(ctx, key, value, valuePredicate, true)
+}
+
+func (s *Store) setWithOptionalPredicate(ctx context.Context, key, value []byte, valuePredicate kv.Predicate, usePredicate bool) error {
+	if len(key) == 0 {
+		return kv.ErrMissingKey
+	}
+	if value == nil {
+		return kv.ErrMissingValue
+	}
+
+	item := DynKVItem{
+		PartKey:   s.getPartitionKey(key),
+		ItemKey:   string(key),
+		ItemValue: string(value),
+	}
+
+	marshaledItem, err := dynamodbattribute.MarshalMap(item)
+	if err != nil {
+		return fmt.Errorf("%s: %w", err, kv.ErrOperationFailed)
```

Comment:
Is every error from any store a `kv.ErrOperationFailed`? What's the point of having this error then?

Reply:
I'm not sure I fully understand your question, so I will try to explain from my POV:
Errors that can be well indetified from **our** `KV` aspect, such as missing key, missing value or failed predicate, has a specialized well defined. Other errors, mainly these coming from the underlying engine/DB - we wrap the error with the general `ErrOperationFailed`.
In the specific case you commented above, this is a marshaling error that may be a result of changing the DynamoDB table structure, without  modifying our code accordingly, or the other way around (maybe due to versions mismatch or similar). This is just one example and there are probably more, but I think the common base for all these is to wrap various failures, mainley the ones that are underlying layer based, with "our own" `kv` error type.
Please let me know if this answers your question

Reply:
I see what you mean, but I'm not sure this is the right scope for that.
I think we will need a better definition of errors we return, and how we handle errors from the underlying infrastructure(s).
Note that the more specific errors you mentioned, such as `ErrMissingArgument` are identifiable by our code, while (at least most of) the ones that are simply wrapped by the very generic `ErrOperationFailed`, are a result of error returned from beneath, and it will require to add our own error handling code for these. Moreover, I'm not sure how actionable it is, as we (at least I) haven't fully gotten into all the errors that can happen.
One last point - the same behavior (more or less) is present in the other stores implementation, and if we revise this one, we will need to align all of them.
How about we handle it in a separate issue, for better error handling and reporting?